File: src/expression.cpp
#include "expression.h"

std::ostream& Expression::format(std::ostream& os) const {
    os << "()";
    return os;
}

std::ostream& operator<<(std::ostream& os, const Expression& expression)
{
    return expression.format(os);
}

BinaryExpression::BinaryExpression(Expression* left, Token op, Expression* right) {
    this->left = left;
    this->op = op;
    this->right = right;
}

std::ostream& BinaryExpression::format(std::ostream& os) const {
    os << "(" << *left << " " << op.string << " " << *right << ")";
    return os;
}

std::ostream& operator<<(std::ostream& os, const BinaryExpression& expression)
{
    return expression.format(os);
}

IntLiteralExpression::IntLiteralExpression(const Token token) {
    this->token = token;
}

std::ostream& IntLiteralExpression::format(std::ostream& os) const {
    os << "(" << token.string << ")";
    return os;
}

std::ostream& operator<<(std::ostream& os, const IntLiteralExpression& expression)
{
    return expression.format(os);
}

StringLiteralExpression::StringLiteralExpression(const Token token) {
    this->token = token;
}

std::ostream& StringLiteralExpression::format(std::ostream& os) const {
    os << "(\"" << token.string << "\")";
    return os;
}

std::ostream& operator<<(std::ostream& os, const StringLiteralExpression& expression)
{
    return expression.format(os);
}

File: gliamc/src/expression.go
package src

import "fmt"

type Expression interface {
	codegen() string
}

type TypeExpression interface {
	codegen() string
}

type BinaryExpression struct {
	Left  Expression
	Op    OperatorType
	Right Expression
}

type NumberLiteralExpression struct {
	Token Token
	Value float64
}

type StringLiteralExpression struct {
	Token Token
}

type BoolLiteralExpression struct {
	Token Token
}

type IdentifierLiteralExpression struct {
	Identifier Token
}

type IdentifierTypeExpression struct {
	Identifier Token
}

func (expr BinaryExpression) Codegen() string {
	output := expr.Left.codegen()

	if expr.Op == OP_PLUS {
		output += " + "
	} else if expr.Op == OP_MINUS {
		output += " - "
	} else if expr.Op == OP_MULT {
		output += " * "
	} else if expr.Op == OP_DIV {
		output += " / "
	}

	return output + expr.Right.codegen()
}

func (expr NumberLiteralExpression) codegen() string {
	return fmt.Sprintf("%f", expr.Value)
}
